optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsJoinBiConstraintStream.java

```diff
@@ -0,0 +1,142 @@
+    private final AbstractBiJoiner<A, B> biJoiner;
+    private final PatternDSL.PatternDef<B> rightPattern;
+
+    public DroolsJoinBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractUniConstraintStream<Solution_, A> parent,
+            DroolsAbstractUniConstraintStream<Solution_, B> otherStream, BiJoiner<A, B> biJoiner) {
+        super(constraintFactory, null);
+        this.leftParentStream = parent;
+        this.rightParentStream = otherStream;
+        this.biJoiner = (AbstractBiJoiner<A, B>) biJoiner;
+        this.rightPattern = otherStream.getPattern().expr(getLeftVariableDeclaration(), (b, a) -> matches(a, b));
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, A> getLeftParentStream() {
+        return leftParentStream;
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, B> getRightParentStream() {
+        return rightParentStream;
+    }
+
+    @Override
+    public Declaration<A> getLeftVariableDeclaration() {
+        return leftParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<A> getLeftPattern() {
+        return leftParentStream.getPattern();
+    }
+
+    @Override
+    public Declaration<B> getRightVariableDeclaration() {
+        return rightParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<B> getRightPattern() {
+        return rightPattern;
+    }
+
+    private static boolean matches(Object left, Object right, JoinerType type) {
+        switch (type) {
+            case EQUAL:
+                return left.equals(right);
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+            case GREATER_THAN:
+            case GREATER_THAN_OR_EQUAL:
+                return matchesComparable((Comparable)left, right, type);
+            case CONTAINING:
+                return ((Collection) left).contains(right);
+            case DISJOINT:
+            case INTERSECTING:
+                return matchesCollection((Collection)left, (Collection) right, type);
+            default:
+                throw new IllegalStateException("Unsupported joiner type (" + type + ").");
+        }
+    }
+
+    private static boolean matchesComparable(Comparable left, Object right, JoinerType type) {
+        int comparison = left.compareTo(right);
+        switch (type) {
+            case LESS_THAN:
+                return comparison < 0;
+            case LESS_THAN_OR_EQUAL:
+                return comparison <= 0;
+            case GREATER_THAN:
+                return comparison > 0;
+            case GREATER_THAN_OR_EQUAL:
+                return comparison >= 0;
+            default:
+                throw new IllegalStateException("Unsupported joiner type (" + type + ").");
+        }
+    }
+
+    private static boolean matchesCollection(Collection left, Collection right, JoinerType type) {
```

Comment:
code style nitpick: would you mind if we name the JoinerType "joinerType" instead of just "type"?
And "left" as "leftCollection" etc?